File: ai-leak-detection/detect_leak.py
import pytesseract
from PIL import Image, UnidentifiedImageError
from rapidfuzz import fuzz
import os

pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'

with open('master_paper.txt', 'r') as f:
    master_text = f.read()

name_to_center = {
    'Ramesh': 'Center_Kanpur',
    'Suresh': 'Center_Lucknow',
    'Mahesh': 'Center_Delhi'
}

refcode_to_center = {
    'KNP7': 'Center_Kanpur',
    'LKO7': 'Center_Lucknow',
    'DEL7': 'Center_Delhi'
}

def check_leak(image_path):
    print(f"\n--- Testing: {image_path} ---")

    if not os.path.exists(image_path):
        print("ERROR: File nahi mili. Filename check karo.")
        return

    try:
        img = Image.open(image_path)
    except UnidentifiedImageError:
        print("ERROR: Ye file ek valid image nahi hai (corrupt ho sakti hai).")
        return

    extracted_text = pytesseract.image_to_string(img)

    if extracted_text.strip() == "":
        print("ERROR: OCR ko koi text nahi mila is image mein.")
        return
    print(f"OCR ne ye nikala: {extracted_text}")
    match_score = fuzz.ratio(extracted_text, master_text)
    print(f"Match score with master paper: {match_score:.1f}%")

    if match_score > 50:
        print("CONFIRMED: Ye leaked paper hamara exam paper hai.")

        found_center = None
        matched_via = None

        for name, center in name_to_center.items():
            if name in extracted_text:
                found_center = center
                matched_via = f"name fingerprint ('{name}')"
                break

        if not found_center:
            for code, center in refcode_to_center.items():
                if code in extracted_text:
                    found_center = center
                    matched_via = f"reference code fingerprint ('{code}')"
                    break

        if found_center:
            print(f"Fingerprint matched via: {matched_via}")
            print(f"Leak traced to: {found_center}")
        else:
            print("Dono fingerprints (naam aur ref code) nahi mile.")
            print("Suggestion: Manual review needed, image bahut zyada crop/blur hai.")
    else:
        print("Ye hamara exam paper nahi lagta (match score bahut kam hai).")

check_leak('leaked_photo.png')
check_leak('blurry_leak.png')
check_leak('random_image.png')

File: ai-leak-detection/detectLeak.js
import fs from 'fs';
import Tesseract from 'tesseract.js';
import stringSimilarity from 'string-similarity';

const masterText = fs.readFileSync('master_paper.txt', 'utf-8');

const nameToCenter = {
  Ramesh: 'Center_Kanpur',
  Suresh: 'Center_Lucknow',
  Mahesh: 'Center_Delhi'
};

const refcodeToCenter = {
  KNP7: 'Center_Kanpur',
  LKO7: 'Center_Lucknow',
  DEL7: 'Center_Delhi'
};

async function checkLeak(imagePath) {
  console.log(`\n--- Testing: ${imagePath} ---`);

  if (!fs.existsSync(imagePath)) {
    console.log('ERROR: File nahi mili. Filename check karo.');
    return;
  }

  let extractedText;
  try {
    const result = await Tesseract.recognize(imagePath, 'eng');
    extractedText = result.data.text;
  } catch (err) {
    console.log('ERROR: Ye file ek valid image nahi hai (corrupt ho sakti hai).');
    return;
  }

  if (extractedText.trim() === '') {
    console.log('ERROR: OCR ko koi text nahi mila is image mein.');
    return;
  }

  const matchScore = stringSimilarity.compareTwoStrings(extractedText, masterText) * 100;
  console.log(`Match score with master paper: ${matchScore.toFixed(1)}%`);

  if (matchScore > 30) {
    console.log('CONFIRMED: Ye leaked paper hamara exam paper hai.');

    let foundCenter = null;
    let matchedVia = null;

    for (const [name, center] of Object.entries(nameToCenter)) {
      if (extractedText.toLowerCase().includes(name.toLowerCase())) {
        foundCenter = center;
        matchedVia = `name fingerprint ('${name}')`;
        break;
      }
    }

    if (!foundCenter) {
      for (const [code, center] of Object.entries(refcodeToCenter)) {
        if (extractedText.toLowerCase().includes(code.toLowerCase())) {
          foundCenter = center;
          matchedVia = `reference code fingerprint ('${code}')`;
          break;
        }
      }
    }

    if (foundCenter) {
      console.log(`Fingerprint matched via: ${matchedVia}`);
      console.log(`Leak traced to: ${foundCenter}`);
    } else {
      console.log('Dono fingerprints (naam aur ref code) nahi mile.');
      console.log('Suggestion: Manual review needed, image bahut zyada crop/blur hai.');
    }
  } else {
    console.log('Ye hamara exam paper nahi lagta (match score bahut kam hai).');
  }
}

async function runAllTests() {
  await checkLeak('leaked_photo.png');
  await checkLeak('blurry_leak.png');
  await checkLeak('random_image.png');
}

runAllTests();